src/ReachSets/reach.jl

```diff
@@ -211,6 +214,130 @@ function reach(S::AbstractSystem,
+                          decomposition)
+- `set_type_init`      -- (optional, default: `Hyperrectangle`) set type for the
+                          approximation of the initial states (during
+                          decomposition)
+- `ε_iter`             -- (optional, default: `Inf`) error bound for the
+                          approximation of the states ``X_k``, ``k>0``
+- `set_type_iter`      -- (optional, default: `Hyperrectangle`) set type for the
+                          approximation of the states ``X_k``, ``k>0``
+- `assume_sparse`      -- (optional, default: `true`) if true, it is assumed
+                          that the coefficients matrix (exponential) is sparse;
+                          otherwise, it is transformed to a full matrix
+- `assume_homogeneous` -- (optional, default: `false`) if true, it is assumed
+                          that the system has no input (linear system), and in
+                          case it has one, the input is ignored
+- `numeric_type`       -- (optional, default: `Float64`) numeric type of the
+                          resulting set
+- `lazy_X0`            -- (optional, default: `false`) if true, transform the
+                          set of initial states to the caretsian product of
+                          two-dimensional polygons; otherwise, the given input,
+                          as a lazy set, is passed to the backend
+- `kwargs`             -- (optional) additional arguments that are passed to the
+                          backend
+
+### Notes
+
+A dictionary with available algorithms is available via
+`Reachability.available_algorithms`.
+"""
+function d_reach(HS::HybridSystem, # Does AbstractSystem a superset of HybridSystem?
+               N::Int;
+               X0::BallInf; # TODO is it correct to way to specify initial set for HS?
+               algorithm::String="explicit",
+               ε_init::Float64=Inf,
+               set_type_init::Type{<:LazySet}=Hyperrectangle,
+               ε_iter::Float64=Inf,
+               set_type_iter::Type{<:LazySet}=Hyperrectangle,
+               assume_sparse=true,
+               assume_homogeneous=false,
+               numeric_type::Type=Float64,
+               lazy_X0=false,
+               kwargs...)::Vector{<:LazySet}
+               # unpack arguments
+               kwargs_dict = Dict(kwargs)
+
+
+
+               # list containing the arguments passed to any reachability function
+               args = []
+               waiting_list = []
+               # coefficients matrix
+               #TODO get start state. For now we assume that it is the first location
+               cur_loc_id = 1
+               cur_loc = HS.modes[init_loc_id]
+
+               #TODO push_initial_set
+               push!((cur_loc[1], X0), waiting_list)
+               i = 0
+               while (!isempty(waiting_list) and i < 15) #TODO add variable for max iteration number
```

Comment:
Is it common to have a number of iterations? It might be helpful for testing, but in general I would rather restrict the number of jumps (which we would have to store in this case).

Reply:
At least SpaceEx has it and it seems to be very useful feature. For instance, if Reachability tool cannot find fixpoint(or it could take ages to calculate it), then people usually set up some number of maximum iterations, which depends on the model.

Reply:
Both versions introduce a termination criterion.
I know some other tools that fix the jump depth. Then you can output the concrete statement that "the bad states are not reachable withing 5 jumps."
On the other hand, by fixing the number of iterations, the output depends on the iteration order (and so it is in general nondeterministic, which is pretty annoying), while fixed jump depth search is deterministic. Consider the standard search strategies [breadth-first search](https://en.wikipedia.org/wiki/Breadth-first_search) vs. [depth-first search](https://en.wikipedia.org/wiki/Depth-first_search). With fixed jump depth you search in the same area of the search tree (only the order changes).

The good thing about fixing the number of iterations is that you get a rather reliable termination criterion (because the search tree may grow exponentially, which makes the jump depth restriction quite useless). Maybe at some point we want to have both, but for now I find your setting more useful for testing. You should, however, add a feedback output in the end that tells you if the analysis was complete or if it was terminated prematurely.

Reply:
I believe in SpaceEx you can choose ```-1```, which means the algorithm terminates only when a fixed point is reached.